src/corehost/corehost.cpp

```diff
@@ -0,0 +1,85 @@
+#include "common/trace.h"
+#include "common/pal.h"
+#include "common/utils.h"
+
+enum StatusCode
+{
+    Failure                  = 0x87FE0000,
+    CoreHostLibLoadFailure    = Failure | 0x1,
+    CoreHostSvcLibLoadFailure = Failure | 0x2,
+};
+
+extern int corehost_main(const int argc, const pal::char_t* argv[]);
+typedef int (*corehost_main_fn) (const int argc, const pal::char_t* argv[]);
+
+corehost_main_fn load_corehost(const pal::string_t& corehost_path, pal::dll_t* h_corehost)
+{
+    if (!pal::load_library(corehost_path.c_str(), h_corehost))
+    {
+        return nullptr;
+    }
+
+    return (corehost_main_fn) pal::get_symbol(*h_corehost, "corehost_main");
+}
+
+void setup_tracing()
+{
+    // Read trace environment variable
+    pal::string_t trace_str;
+    if (!pal::getenv(_X("COREHOST_TRACE"), &trace_str))
+    {
+        return;
+    }
+
+    auto trace_val = pal::xtoi(trace_str.c_str());
+    if (trace_val > 0)
+    {
+        trace::enable();
+        trace::info(_X("Tracing enabled"));
+    }
+}
+
+#if defined(_WIN32)
+int __cdecl wmain(const int argc, const pal::char_t* argv[])
+#else
+int main(const int argc, const pal::char_t* argv[])
+#endif
+{
+    setup_tracing();
+
+    pal::string_t env_str;
+    pal::dll_t corehost;
+
+    // First probe for the env variable. If set and missing, flag fatal error.
+    // This is to load custom hosts: corerun or clihost.
+    if (pal::getenv(_X("COREHOST_LIBPATH"), &env_str))
+    {
+        corehost_main_fn host_main = load_corehost(env_str, &corehost);
```

Comment:
Please distinguish between inability to load the binary versus not finding the exported function. 
This will make the load failures more diagnosable.